src/lib.rs

```diff
@@ -715,251 +738,307 @@ enum Decoder {
+                [Value::U8(hi), Value::U8(lo)] => {
+                    Cow::Owned(Value::U16(u16::from_be_bytes([*hi, *lo])))
+                }
                 _ => panic!("U16Be: expected (U8, U8)"),
             },
             Expr::U16Le(bytes) => match bytes.eval_value(scope).unwrap_tuple().as_slice() {
-                [Value::U8(lo), Value::U8(hi)] => Value::U16(u16::from_le_bytes([*lo, *hi])),
+                [Value::U8(lo), Value::U8(hi)] => {
+                    Cow::Owned(Value::U16(u16::from_le_bytes([*lo, *hi])))
+                }
                 _ => panic!("U16Le: expected (U8, U8)"),
             },
             Expr::U32Be(bytes) => match bytes.eval_value(scope).unwrap_tuple().as_slice() {
                 [Value::U8(a), Value::U8(b), Value::U8(c), Value::U8(d)] => {
-                    Value::U32(u32::from_be_bytes([*a, *b, *c, *d]))
+                    Cow::Owned(Value::U32(u32::from_be_bytes([*a, *b, *c, *d])))
                 }
                 _ => panic!("U32Be: expected (U8, U8, U8, U8)"),
             },
             Expr::U32Le(bytes) => match bytes.eval_value(scope).unwrap_tuple().as_slice() {
                 [Value::U8(a), Value::U8(b), Value::U8(c), Value::U8(d)] => {
-                    Value::U32(u32::from_le_bytes([*a, *b, *c, *d]))
+                    Cow::Owned(Value::U32(u32::from_le_bytes([*a, *b, *c, *d])))
                 }
                 _ => panic!("U32Le: expected (U8, U8, U8, U8)"),
             },
-            Expr::SeqLength(seq) => match seq.eval(scope) {
+            Expr::SeqLength(seq) => match &*seq.eval(scope) {
                 Value::Seq(values) => {
                     let len = values.len();
-                    Value::U32(len as u32)
+                    Cow::Owned(Value::U32(len as u32))
                 }
                 _ => panic!("SeqLength: expected Seq"),
             },
-            Expr::SubSeq(seq, start, length) => match seq.eval(scope) {
+            Expr::SubSeq(seq, start, length) => match seq.eval(scope).into_owned() {
                 Value::Seq(values) => {
                     let start = start.eval_value(scope).unwrap_usize();
                     let length = length.eval_value(scope).unwrap_usize();
                     let values = &values[start..];
                     let values = &values[..length];
-                    Value::Seq(values.to_vec())
+                    Cow::Owned(Value::Seq(values.to_vec()))
                 }
                 _ => panic!("SubSeq: expected Seq"),
             },
             Expr::FlatMap(expr, seq) => match expr.as_ref() {
-                Expr::Lambda(name, expr) => match seq.eval(scope) {
+                Expr::Lambda(name, expr) => match seq.eval(scope).into_owned() {
                     Value::Seq(values) => {
                         let mut vs = Vec::new();
                         for v in values {
                             scope.push(name.clone(), v);
```

Comment:
Here is a prime example of where `Cow<'static, str>` would save us allocation costs.

Reply:
That's a good point, I'll investigate in a future branch.